### PR TITLE
Doc fragments to plugins

### PR DESCRIPTION
**Pull request by bcoca**
promote doc fragments to full fledged plugins

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

doc fragments

#TODO:
 - create/force base class?
 - subfolders per plugin type? (not just modules use these)

### CODE REVIEW

File: test/runner/lib/sanity/ansible_doc.py [hunk 1]
```diff
@@ -41,9 +41,11 @@ def test(self, args, targets, python_version):
         skip_file = 'test/sanity/ansible-doc/skip.txt'
         skip_modules = set(read_lines_without_comments(skip_file, remove_blank_lines=True))
 
+        # TODO: use documentable plugins from contsants instead
```
Comment (mattclay) [inline test/runner/lib/sanity/ansible_doc.py]: Typo:

```suggestion
        # TODO: use documentable plugins from constants instead
```

**Review by mattclay — COMMENTED** [review]
Also need to update: https://github.com/ansible/ansible/blob/fd56788a142efde6b24553978906d43d2d4ea771/test/runner/lib/classification.py#L520

**Comment by mattclay** [comment]
@bcoca Should this PR be marked `WIP`?

**Comment by mattclay** [comment]
Another path to update:

https://github.com/ansible/ansible/blob/fd56788a142efde6b24553978906d43d2d4ea771/test/runner/lib/sanity/yamllint.py#L47

**Comment by bcoca** [comment]
@mattclay nope, i just didnt realize soo many things hardcoded the path to doc fragments

**Review by gundalow — CHANGES_REQUESTED** [review]
For my understanding, what are the benefits of this over the current implementation? 

1) `docs/rst/` has lots of references to the fragments directory which need updating in this PR

2) I wonder of this neecsyadfing to the developer section of the Porting guide? 

3) Once merged this needs announcing https://github.com/ansible/community/issues/346

4) After merge need to review all PRs which add/update fragments

**Comment by bcoca** [comment]
Two main changes, we put doc_fragments under the same umbrella as other plugins (easier to deal with less exceptions) AND we allow for user provided doc_fragments, currently module/plugin authors must repeat all their docs when providing modules or are required to commit to ansible/ansible to make them available.

1. I'll have to update those
2. Yes
3. of course
4. now it makes sense, previous phrasing confused me, those will be in conflict since i moved the file

File: lib/ansible/config/base.yml [hunk 2]
```diff
@@ -352,6 +352,14 @@ LOCALHOST_WARNING:
   - {key: localhost_warning, section: defaults}
   type: boolean
   version_added: "2.6"
+DOC_FRAGMENT_PLUGIN_PATH:
+  name: documentation fragment plugins path
+  default: ~/.ansible/plugins/doc_fragments:/usr/share/ansible/plugins/doc_fragments
+  description: Colon separated paths in which Ansible will search for Documentation Fragments Plugins.
+  env: [{name: ANSIBLE_DOC_FRAGMENT_PLUGINS}]
```
Comment (webknjaz) [inline lib/ansible/config/base.yml:359]: Could you please use native YAML syntax and do multi-line?

File: lib/ansible/config/base.yml [hunk 3]
```diff
@@ -352,6 +352,14 @@ LOCALHOST_WARNING:
...
+  ini:
+  - {key: doc_fragment_plugins, section: defaults}
```
Comment (webknjaz) [inline lib/ansible/config/base.yml:361]: Multi-line would look better here as well

**Comment by bcoca** [comment]
bot_status

**Review by nitzmahone — APPROVED** [review]

**Pull request merged**